internal/provider/stack_migrate_resource.go

```diff
@@ -0,0 +1,169 @@
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to open stacks configuration",
+			"An error occurred while trying to open the stacks configuration: "+err.Error())
+		return
+	}
+
+	dependencyLocksHandle, closeLock, err := r.stateOps.OpenDependencyLockFile(sourceBundleHandle, "./.terraform.lock.hcl")
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to open dependency lock file",
+			"An error occurred while trying to open the dependency lock file: "+err.Error())
+		return
+	}
+
+	providerCacheHandle, closeProviderCache, err := r.stateOps.OpenProviderCache("./.terraform/providers/")
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to open provider cache",
+			"An error occurred while trying to open the provider cache: "+err.Error())
+		return
+	}
+
+	events, err := r.stateOps.MigrateTFState(
+		tfStateHandle,
+		stackConfigHandle,
+		dependencyLocksHandle,
+		providerCacheHandle,
+		map[string]string{}, // resources map, to be filled as needed
+		map[string]string{}, // modules map, to be filled as needed
+	)
+	if err != nil {
+		resp.Diagnostics.AddError("Unable to migrate Terraform state",
+			"An error occurred while trying to migrate the Terraform state: "+err.Error())
+		return
+	}
+
+	for {
+		item, err := events.Recv()
+		if err == io.EOF {
+			break // all done
+		} else if err != nil {
+			resp.Diagnostics.AddError("Error receiving migration events",
+				"An error occurred while receiving migration events: "+err.Error())
+			return
+		}
+
+		switch result := item.Result.(type) {
+		case *stacks.MigrateTerraformState_Event_AppliedChange:
+			for _, change := range result.AppliedChange.Descriptions {
+				fmt.Println(jsonOpts.Format(change)) // Handle
+			}
+		case *stacks.MigrateTerraformState_Event_Diagnostic:
+			// Handle diagnostics
+		}
+	}
+
+	// Close all handles, refactor to use defer if required
+	if err := closeTFState(); err != nil {
+		resp.Diagnostics.AddError("Error closing Terraform state handle",
+			"An error occurred while closing the Terraform state handle: "+err.Error())
+		return
+	}
+	if err := closeSourceBundle(); err != nil {
+		resp.Diagnostics.AddError("Error closing source bundle handle",
+			"An error occurred while closing the source bundle handle: "+err.Error())
+		return
+	}
+	if err := closeConfig(); err != nil {
+		resp.Diagnostics.AddError("Error closing stacks configuration handle",
+			"An error occurred while closing the stacks configuration handle: "+err.Error())
+		return
+	}
+	if err := closeLock(); err != nil {
+		resp.Diagnostics.AddError("Error closing dependency lock file handle",
+			"An error occurred while closing the dependency lock file handle: "+err.Error())
+		return
+	}
+	if err := closeProviderCache(); err != nil {
+		resp.Diagnostics.AddError("Error closing provider cache handle",
+			"An error occurred while closing the provider cache handle: "+err.Error())
+		return
+	}
```

Comment:
optional - you could do these in defer statements so that they are properly closed even if we return early. But, I think as the main client is being closed anyway this isn't as important since everything held in memory will be freed when the process exits.